sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -17,19 +17,130 @@
 
 package org.apache.spark.sql.execution.datasources.v2
 
-import org.apache.spark.sql.catalyst.expressions.{And, Expression, NamedExpression, ProjectionOverSchema, SubqueryExpression}
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
 import org.apache.spark.sql.catalyst.planning.ScanOperation
-import org.apache.spark.sql.catalyst.plans.logical.{Filter, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, Filter, LogicalPlan, Project}
 import org.apache.spark.sql.catalyst.rules.Rule
-import org.apache.spark.sql.connector.read.{Scan, V1Scan}
+import org.apache.spark.sql.catalyst.util.toPrettySQL
+import org.apache.spark.sql.connector.read.{Scan, SupportsPushDownAggregates, V1Scan}
 import org.apache.spark.sql.execution.datasources.DataSourceStrategy
 import org.apache.spark.sql.sources
+import org.apache.spark.sql.sources.{AggregateFunc, Aggregation}
 import org.apache.spark.sql.types.StructType
 
-object V2ScanRelationPushDown extends Rule[LogicalPlan] {
+object V2ScanRelationPushDown extends Rule[LogicalPlan] with AliasHelper {
```

Comment:
We need to rethink how to do v2 scan pushdown. The current code is really hacky and hard to maintain.

An optimized plan usually looks like
```
Aggregate
  Project
    Filter
      Scan
```

The algorithm can be
1. we traverse the query plan tree bottom-up, starting with the v2 scan node.
2. If we hit a Filter, apply filter pushdown, and keep going
3. if we hit a Project, collect the alias map and keep going
4. if we hit a Aggregate, apply the alias map to get the actual grouping/aggregate expression, and apply aggregate pushdown, and build the `Scan`
5. If we hit other nodes, apply column prunin and build the `Scan`

Reply:
I will redo this

Reply:
@cloud-fan Sorry, a dumb question here: if i have a simple aggregate such as `SELECT Min(c1) WHERE c2 > 1`, when I traverse up the tree, I hit V2 Scan node first, then Filter, then Project, then Aggregate. Since I don't have other nodes, seems to me I will exit the traverse without reaching step 5?

Reply:
It goes without saying that we need to stop if there are no more nodes...

Reply:
Sorry to not be clear enough:
5. If we hit other nodes or there is no more node, apply column prunin and build the Scan